File: csharp/src/Audio/AudioGraph.cs
using System;
using System.Collections.Generic;

namespace BQt.Audio
{
    public class AudioGraph
    {
        private readonly Dictionary<string, IAudioNode> _nodes = new Dictionary<string, IAudioNode>();
        private readonly Dictionary<string, List<string>> _links = new Dictionary<string, List<string>>();
        private readonly object _lock = new object();
        private bool _isRunning = false;

        public void AddNode(IAudioNode node)
        {
            lock (_lock)
            {
                _nodes[node.Name] = node;
            }
        }

        public void Connect(string srcName, string destName)
        {
            lock (_lock)
            {
                if (!_nodes.ContainsKey(srcName))
                    throw new ArgumentException($"node {srcName} not found");
                if (!_nodes.ContainsKey(destName))
                    throw new ArgumentException($"node {destName} not found");

                if (!_links.ContainsKey(srcName))
                    _links[srcName] = new List<string>();

                _links[srcName].Add(destName);
            }
        }

        public void ProcessBlock(float[] buffer)
        {
            lock (_lock)
            {
                foreach (var node in _nodes.Values)
                {
                    node.Process(buffer);
                }
            }
        }

        public void Start()
        {
            lock (_lock)
            {
                if (_isRunning) return;
                _isRunning = true;
                Console.WriteLine("AudioGraph: Audio processing thread STARTED. (Simulating JUCE DeviceManager)");
            }
        }

        public void Stop()
        {
            lock (_lock)
            {
                if (!_isRunning) return;
                _isRunning = false;
                Console.WriteLine("AudioGraph: Audio processing thread STOPPED.");
            }
        }
    }
}

File: rust/src/audio/audio_graph.rs
use std::collections::HashMap;



pub trait AudioNode: Send + Sync {
    fn name(&self) -> &str;
    fn process(&mut self, buffer: &mut [f32]);
}

pub struct AudioGraph {
    nodes: HashMap<String, Box<dyn AudioNode>>,
    links: HashMap<String, Vec<String>>,
    is_running: bool,
}

impl AudioGraph {
    pub fn new() -> Self {
        Self {
            nodes: HashMap::new(),
            links: HashMap::new(),
            is_running: false,
        }
    }

    pub fn add_node(&mut self, node: Box<dyn AudioNode>) {
        self.nodes.insert(node.name().to_string(), node);
    }

    pub fn connect(&mut self, src_name: &str, dest_name: &str) -> Result<(), String> {
        if !self.nodes.contains_key(src_name) {
            return Err(format!("node {} not found", src_name));
        }
        if !self.nodes.contains_key(dest_name) {
            return Err(format!("node {} not found", dest_name));
        }

        self.links.entry(src_name.to_string())
            .or_insert_with(Vec::new)
            .push(dest_name.to_string());

        Ok(())
    }

    pub fn process_block(&mut self, buffer: &mut [f32]) {
        for node in self.nodes.values_mut() {
            node.process(buffer);
        }
    }

    pub fn start(&mut self) {
        if self.is_running {
            return;
        }
        self.is_running = true;
        println!("AudioGraph: Audio processing thread STARTED. (Simulating JUCE DeviceManager)");
    }

    pub fn stop(&mut self) {
        if !self.is_running {
            return;
        }
        self.is_running = false;
        println!("AudioGraph: Audio processing thread STOPPED.");
    }
}
